src/main/java/org/digitalstack/chess/ChessBoard.java

```diff
@@ -12,10 +14,33 @@ public ChessBoard() {
     }
 
     public void add(Pawn pawn, int xCoordinate, int yCoordinate, PieceColor pieceColor) {
-        throw new UnsupportedOperationException("Need to implement ChessBoard.add()");
+
+            int count = 0;
+            for(int i = 0; i < 7; i++ )
+                for(int j = 0; j < 7; j++ )
+                    if(pieces[i][j] != null )
+                        count ++;
+            if(isLegalBoardPosition(xCoordinate, yCoordinate) && count < 7){
```

Comment:
Negate this condition and fail fast. You will remove one layer of imbrication this way.